src/WooCommerce.php

```diff
@@ -128,6 +128,148 @@ public static function woocommerce_single_product_image_thumbnail_html($html, $t
     return preg_replace('/(<a\s+)/i', '<a data-srcset="' . $srcset . '" data-sizes="' . $srcsizes . '" ', $html);
   }
 
+  /**
+   * Adds additional images from product variation gallery.
+   *
+   * @implements woocommerce_product_get_gallery_image_ids
+   */
+  public static function woocommerce_product_get_gallery_image_ids($value, $obj) {
+    global $product;
+
+    if (!$product || $product->get_type() !== 'variable') {
+      return $value;
+    }
+
+    $variations = $product->get_available_variations();
+    $images = [];
+    // Add images from variation gallery.
+    foreach ($variations as $variation) {
+      $images = array_merge(
+        $images,
+        static::getVariationImages($variation)
+      );
+    }
+
+    return array_merge($value, $images);
+  }
+
+  /**
+   * Collects all related images URLs of a variable product variations.
+   *
+   * If current product has variations, we collect the URLs for its featured
+   * image, the general gallery images and the images in each of its variations
+   * galleries. We print this data as a JS object.
+   *
+   * When the user selects a variation in the frontend UI, we hide all images in
+   * the product image gallery slider, except for the product features image,
+   * the general gallery images and the selected variation gallery images (if
+   * those exist).
+   */
+  public static function woocommerce_product_thumbnails() {
+    global $product;
+
+    if ($product->get_type() !== 'variable') {
+      return;
+    }
+
+    // Get size of images for the produt thumbnails slider.
+    $gallery_thumbnail = wc_get_image_size('gallery_thumbnail');
+    $thumbnail_size = apply_filters(
+      'woocommerce_gallery_thumbnail_size',
+      [$gallery_thumbnail['width'], $gallery_thumbnail['height']]
+    );
+
+    $variation_images = [];
+    // Collect product main image and general images gallery.
+    $variation_images['gallery'] = [
+      wp_get_attachment_image_src(get_post_thumbnail_id(
+        $product->get_id()),
+        $thumbnail_size
+      )[0],
+    ];
+    foreach ($product->get_gallery_image_ids('edit') as $image_id) {
+      $variation_images['gallery'][] = wp_get_attachment_image_src(
+        $image_id,
+        $thumbnail_size
+      )[0];
+    }
+
+    // Collect product variations images.
+    $variations = $product->get_available_variations();
+    foreach ($variations as $variation) {
+      $variation_images[$variation['variation_id']] =
+        static::getVariationImagesUrls($variation, $thumbnail_size);
+    }
+
+    echo sprintf(
+      '<script>var product_variation_images = %s;</script>',
+      json_encode($variation_images)
+    );
```

Comment:
In general in this PR - not a fan of the excessive line wrapping…
```suggestion
    echo sprintf('<script>var product_variation_images = %s;</script>', json_encode($variation_images));
```